### PR TITLE
Add tests for data objects stuck in locked status, prevent nullptr dereference (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses irods/irods#6154
Addresses irods/irods#6479
Addresses irods/irods#6880

Companion PR: https://github.com/irods/irods/pull/6884

Running tests now.

### CODE REVIEW

File: packaging/resource_suite_s3_cache.py [hunk 1]
```diff
@@ -594,6 +594,95 @@ def test_multiple_s3_endpoints_replication_issue1858(self):
...
+            self.user0.assert_icommand(['iput', '-f', physical_path, logical_path])
+            self.assertEqual(str(1), lib.get_replica_status(self.user0, filename, replica_number_in_s3))
+
+        finally:
+            # Set the replica status here so that we can remove the object even if it is stuck in the locked status.
+            self.admin.run_icommand([
+                'iadmin', 'modrepl',
+                'logical_path', logical_path,
+                'replica_number', str(replica_number_in_s3),
+                'DATA_REPL_STATUS', '0'])
```
Comment (alanking) [inline packaging/resource_suite_s3_cache.py:682]: This must be done for all replicas, not just the one in S3.

File: packaging/resource_suite_s3_cache.py [hunk 2]
```diff
@@ -594,6 +594,95 @@ def test_multiple_s3_endpoints_replication_issue1858(self):
...
+            # Ensure that the replica actually contains the contents streamed into it.
+            self.user0.assert_icommand(['iget', logical_path, '-'], 'STDOUT', content)
+
+        finally:
+            # Set the replica status here so that we can remove the object even if it is stuck in the locked status.
+            self.admin.run_icommand([
+                'iadmin', 'modrepl',
+                'logical_path', logical_path,
+                'replica_number', str(replica_number_in_s3),
+                'DATA_REPL_STATUS', '0'])
```
Comment (alanking) [inline packaging/resource_suite_s3_cache.py:640]: This must be done for all replicas, not just the one in S3.

File: packaging/resource_suite_s3_cache.py [hunk 3]
```diff
@@ -594,6 +594,95 @@ def test_multiple_s3_endpoints_replication_issue1858(self):
...
+            # Ensure that the replica actually exists.
+            self.user0.assert_icommand(['iget', logical_path, '-'])
+
+        finally:
+            # Set the replica status here so that we can remove the object even if it is stuck in the locked status.
+            self.admin.run_icommand([
+                'iadmin', 'modrepl',
+                'logical_path', logical_path,
+                'replica_number', str(replica_number_in_s3),
+                'DATA_REPL_STATUS', '0'])
```
Comment (alanking) [inline packaging/resource_suite_s3_cache.py:616]: This must be done for all replicas, not just the one in S3.

**Review by alanking — COMMENTED** [review]
Forgot there are multiple replicas in the cache test.

Also, I guess I have to switch to using tabs for this repo. I will not do that for the tests because the whitespace has to be consistent in python.

**Comment by korydraughn** [comment]
> Also, I guess I have to switch to using tabs for this repo. I will not do that for the tests because the whitespace has to be consistent in python.

Does clang-format touch python files? I wouldn't expect it to.

**Comment by alanking** [comment]
It doesn't, but I know that's the direction we are trying to move in. I was just noting that I'll put the tabs in the .cpp file, but not the python files.

**Comment by alanking** [comment]
Due to irods/irods#6886, the `iput` test for the compound resource suite is failing. Will need to fix that before proceeding

File: packaging/resource_suite_s3_cache.py (hunk 1 above)
Comment (alanking) [reply packaging/resource_suite_s3_cache.py]: Done

File: packaging/resource_suite_s3_cache.py (hunk 2 above)
Comment (alanking) [reply packaging/resource_suite_s3_cache.py]: Done

File: packaging/resource_suite_s3_cache.py (hunk 3 above)
Comment (alanking) [reply packaging/resource_suite_s3_cache.py]: Done

**Comment by alanking** [comment]
Ignoring further complaints from clang-format. I didn't touch the line it currently has an issue with, so it can be fixed whenever we fully switch to tabs.

All plugin tests passed. Waiting for core tests on the iRODS side now

**Comment by alanking** [comment]
iRODS tests are passing, so any further changes to iRODS will be from code review comments in the accompanying PR.

Tagging #2095 as being addressed here now as well

File: packaging/resource_suite_s3_cache.py [hunk 4]
```diff
@@ -105,6 +105,9 @@ def setUp(self):
 
         s3params=os.environ.get('S3PARAMS', s3params);
 
+        self.archive_resc = 's3'
+        self.cache_resc = 'cacheResc'
```
Comment (korydraughn) [inline packaging/resource_suite_s3_cache.py:109]: What are these for? I don't see them being referenced in the file anywhere.

**Review by korydraughn — COMMENTED** [review]
Everything looks good. Just got one question.

File: packaging/resource_suite_s3_cache.py (hunk 4 above)
Comment (alanking) [reply packaging/resource_suite_s3_cache.py:109]: Ah, some leftovers. Will remove. Had some big dreams for some new `lib.py` functions and decided against it for the sake of expediency :)

Comment (korydraughn) [reply packaging/resource_suite_s3_cache.py:109]: Got it. Looking forward to those enhancements for lib.py.

Comment (alanking) [reply packaging/resource_suite_s3_cache.py]: Done

**Review by korydraughn — APPROVED** [review]

**Comment by alanking** [comment]
Added some new tests to confirm that replication and copy APIs behave with invalid keys. All tests are passing.

**Comment by korydraughn** [comment]
Squash to taste and we'll do one last pass over it.

**Comment by alanking** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
Looks good to me.

Once @JustinKyleJames approves, you can go on and merge.

**Comment by JustinKyleJames** [comment]
> Looks good to me.
> 
> Once @JustinKyleJames approves, you can go on and add the pounds and merge it.

I looked this over.  It all looks good / makes sense to me.

**Comment by alanking** [comment]
Okey dokey. #'d and mergin

**Pull request merged**